### PR TITLE
Fail if CJS is detected in FFI

### PR DESCRIPTION
**Pull request by sigma-andex**
**Fail if CJS is detected in FFI, do not fail if FFI js parsing fails**

This is the first change according to the proposal to [remove language-javascript](https://github.com/purescript/purescript/issues/4227) and the proposal to [only support es modules](https://github.com/purescript/purescript/issues/4230).  See also the discussion in the [previous PR within the es modules working group](https://github.com/working-group-purescript-es/purescript/pull/8).
This change addresses the following:
* building fails if CJS is detected in the FFI
* building does not emit an error or warning if the `language-javascript` parsing fails alltogether.
  This is because `language-javascript` does not support certain valid constructs like e.g. dynamic imports or jsx. 

**Note:** One failing bundle test has been removed as `purs bundle` is getting removed anyways.

---

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [x] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Comment by gabejohnson** [comment]
@sigma-andex could you add a link to discussions where this behavior is agreed upon. It would be an enormous breaking change and I don't see mention of it in the [linked issue](https://github.com/purescript/purescript/issues/4227).

**Comment by sigma-andex** [comment]
> @sigma-andex could you add a link to discussions where this behavior is agreed upon. It would be an enormous breaking change and I don't see mention of it in the [linked issue](https://github.com/purescript/purescript/issues/4227).

Updated the description with the link that @thomashoneyman posted and also the link to the discussion in the [previous pr](https://github.com/working-group-purescript-es/purescript/pull/8).

**Comment by JordanMartinez** [comment]
> If people have questions about decisions the working group made we should point them to the proposals and RFCs, as otherwise it could sound like a small group made these decisions without the community participating.

@thomashoneyman Yeah, I wasn't trying to say a small group of people had discussed it and the decision final. Merely that that is where a lot of discussion had taken place. I was expecting to continue any such discussion in this PR if we didn't have any linked issues. But thanks for stating this explicitly as I don't want that perception to be communicated.

File: CHANGELOG.d/breaking_fail-on-cjs-detected.md
Comment (JordanMartinez) [inline CHANGELOG.d/breaking_fail-on-cjs-detected.md:7]: Two comments here. First (and just for future PRs), if this is supposed to be two separate entries, then it should be two separate files.

Second, I think this is actually one entry. Perhaps this rendering summarizes things better? I'd like to hear other core team members' feedback on this point.
```markdown
* If FFI parsing succeeds & CommonJS is detected, fail; otherwise, do not error or warn

  Previously, the compiler would emit an error if it failed to parse the FFI JavaScript file.
  Since the underlying JavaScript parser (i.e. `language-javascript`) fails to parse even
  valid JavaScript files, we cannot consider every failed parse to mean invalid JS files.
  Fixing the parser would require a lot of effort, so we are planning to remove it instead
  in `v0.16.x`.
  
  If the parse succeeds and a CommonJS module is detected, a compiler error is now emitted.
  If the parse fails, we no longer emit a compiler error. While we could emit a warning,
  such a warning will quickly become annoying for FFI files that trigger the buggy paths
  of `language-javascript`. Moreover, we presume that all will be migrating their code to
  ES modules now that CommonJS is being deprecated in the larger JavaScript ecosystem.
```

> CommonJS is being deprecated in the larger JavaScript ecosystem

Regarding this point in the above entry, @sigma-andex could you link to the post that mentioned this? I can't recall where it is.

File: src/Language/PureScript/Make/Actions.hs
Comment (JordanMartinez) [inline src/Language/PureScript/Make/Actions.hs:347]: Looks like this [didn't used to have a `throwError` in front of it](https://github.com/purescript/purescript/pull/4250/files#diff-889762f9d56ac6e7b5dcb3db0b2fca495526bde7fd588e07fdcacaa7d4aca1feL355). The `throwError` was moved into `errorParsingModule`. Would you mind undoing that? I think this change isn't needed for this PR, but correct me if I'm wrong.

**Review by JordanMartinez — COMMENTED** [review]
I think this PR is ready to be approved, but due to the comment from Gabe, I'm going to hold off until getting a response.

File: src/Language/PureScript/Make/Actions.hs
Comment (sigma-andex) [reply src/Language/PureScript/Make/Actions.hs:347]: Gotcha!

Comment (sigma-andex) [reply src/Language/PureScript/Make/Actions.hs:347]: Hmm so the reason is that I'm using `errorParsingModule` also in line [336](https://github.com/purescript/purescript/pull/4250/files#diff-889762f9d56ac6e7b5dcb3db0b2fca495526bde7fd588e07fdcacaa7d4aca1feR336). I could keep it as I was but then I would need to handle that line differently, don't know if that makes much sense

File: CHANGELOG.d/breaking_fail-on-cjs-detected.md
Comment (sigma-andex) [reply CHANGELOG.d/breaking_fail-on-cjs-detected.md]: I think you meant [this discussion](https://github.com/sindresorhus/meta/discussions/15). And there are many libraries that follow this, like e.g. [`node-fetch`](https://github.com/node-fetch/node-fetch#commonjs).

Comment (sigma-andex) [reply CHANGELOG.d/breaking_fail-on-cjs-detected.md]: Should I add this to the changelog?

File: src/Language/PureScript/Make/Actions.hs
Comment (JordanMartinez) [reply src/Language/PureScript/Make/Actions.hs:347]: Ah, ok. Nevermind then.

File: src/Language/PureScript/Make/Actions.hs
Comment (rhendric) [inline src/Language/PureScript/Make/Actions.hs:337]:
```suggestion
  traverse checkFFI parseResult
```

**Review by rhendric — APPROVED** [review]
LGTM (with one trivial suggestion)

File: src/Language/PureScript/Make/Actions.hs
Comment (sigma-andex) [reply src/Language/PureScript/Make/Actions.hs]: Nice!

File: src/Language/PureScript/Errors.hs
Comment (i-am-the-slime) [inline src/Language/PureScript/Errors.hs:716]:
```suggestion
            , line "CommonJS foreign modules (imports using `require`) are no longer supported, please use `import` instead."
```

Comment (i-am-the-slime) [reply src/Language/PureScript/Errors.hs:716]: Or something like this could be beneficial to people not familiar with the term CommonJS and seems more immediately actionable to me.

Comment (sigma-andex) [reply src/Language/PureScript/Errors.hs:716]: I think this is a valid and useful change. However changing the error message always requires updating the tests. So before I do this, I would like to ask here if you all agree on this or have more proposals on how to improve this error message. Thx :-)

Comment (JordanMartinez) [reply src/Language/PureScript/Errors.hs:716]: > Or something like this could be beneficial to people not familiar with the term CommonJS and seems more immediately actionable to me.

Isn't that what the [documentation page for DeprecatedFFICommonJsModule](https://github.com/purescript/documentation/blob/master/errors/DeprecatedFFICommonJSModule.md) is for?

I don't think this change needs to be done.

File: CHANGELOG.d/breaking_fail-on-cjs-detected.md
Comment (JordanMartinez) [reply CHANGELOG.d/breaking_fail-on-cjs-detected.md]: I don't think that was the post I remember you mentioning elsewhere. But since I can't find that post, yeah, let's link to that discussion.

File: src/Language/PureScript/Errors.hs
Comment (natefaubion) [reply src/Language/PureScript/Errors.hs:716]: I think it's worth mentioning what should be used instead, but `imports using require`  isn't the whole story (what about exports using the exports/module object?). I think we should say:

> CommonJS foreign modules are no longer supported. Use native JavaScript/ECMAScript module syntax instead.

**Review by JordanMartinez — APPROVED** [review]

**Comment by gabejohnson** [comment]
@sigma-andex thanks for adding the links!

> Just to clarify, what is "it" in this context? Much of the conversation we've had about the ES modules work has been done via weekly Discord calls.

@JordanMartinez "it" refers to the behavior described in the title "Fail if CJS is detected in FFI". I'm not objecting to the change. Just wanted to see where that decision had been made.

File: src/Language/PureScript/Errors.hs
Comment (sigma-andex) [reply src/Language/PureScript/Errors.hs]: Ok I updated it to @natefaubion 's proposal.

**Comment by JordanMartinez** [comment]
@gabejohnson Ah, thanks for clarifying!

**Pull request merged**
